### ISSUE TITLE
Using Shared library by TA

### ISSUE THREAD

**Issue by YaelAv** [OP-TEE/optee_os#6660]
Is the shared library utilized by the TA in a manner similar to its usage in the Normal World? Specifically, does the OP-TEE OS dynamically load the entire required shared libraries into the address space of the Trusted Application, or is only the required part from the Data section copied? Also, is it considered as a TA? In other words, do I need to use TA to TA calls to access a function from my TA in a shared library?

**Comment by jforissier** [comment]
Hello @YaelAv,

> Is the shared library utilized by the TA in a manner similar to its usage in the Normal World?

Mostly, yes.

> Specifically, does the OP-TEE OS dynamically load the entire required shared libraries into the address space of the Trusted Application,

Yes

> or is only the required part from the Data section copied? 

No, all the PT_LOAD segments are loaded, they contain several sections such as `.data`, `.rodata`, `.text` etc.

> Also, is it considered as a TA?

No, although each shared library has a UUID like a TA and is authenticated (and optionally encrypted) in the same way.

> In other words, do I need to use TA to TA calls to access a function from my TA in a shared library?

Absolutely not. From an application perspective, shared libraries make almost no difference compared with static libraries. You can call the functions directly.

**Comment by YaelAv** [comment]
Thank you for your response. 
To summarize the distinctions between a shared library in the Normal World and a shared library in the Secured World:
1. Memory Aspects: When a TA uses a shared library, it results in a larger memory footprint as all `PT_LOAD` segments, including read-only parts (`.rodata`, `.text`), are shared. However, when a shared library is utilized by an application in the Normal World, only the non-read-only segments are shared.
2. Authorization of the Shared Library: In the Secured World, authorization of the shared library is determined by its UUID taken from the generated .ta file. This is a requirement specific to the Secured World and is not necessary for a shared library in the Normal World.

**Comment by jforissier** [comment]
> Thank you for your response. To summarize the distinctions between a shared library in the Normal World and a shared library in the Secured World:
> 
>     1. Memory Aspects: When a TA uses a shared library, it results in a larger memory footprint as all `PT_LOAD` segments, including read-only parts (`.rodata`, `.text`), are shared. However, when a shared library is utilized by an application in the Normal World, only the non-read-only segments are shared.

Actually, segment sharing is not very different between normal world and secure world. OP-TEE also can share the **read-only** segments of a shared library when several TAs use them. The non-read-only segment can never be shared (NW or SW) because they are susceptible to change.

>     2. Authorization (and optionally encryption) of the Shared Library: In the Secured World, authorization of the shared library is determined by its UUID taken from the generated .ta file. This is a requirement specific to the Secured World and is not necessary for a shared library in the Normal World.

Correct.

**Comment by YaelAv** [comment]
Thanks!

**Comment by YaelAv** [comment]
After testing it, this is what I've got: My TA has two instances. Each instance is supposed to load my shared library using this shared lib UUID ta and elf files. However, I've noticed on the Secure World console that each instance loads the shared library to a unique memory address. This implies that the shared library is being loaded twice. How can I configure OP-TEE to load the shared library just once, ensuring it's at the same memory address for all my TA instances?
